### PR TITLE
Eu-data implementation part 1

### PR DESCRIPTION
**Pull request by niklaslong**
breaking down #162 into smaller PRs

This PR implements the code necessary to download and process CSV for EU data (http://discomap.eea.europa.eu/map/fme/AirQualityUTDExport.htm). A follow-up PR will be needed to implement full list of pollutants for each country, for now we're only using the pollutants implemented by all countries (4/6). 

OpenAQ will still be providing all the data after this PR is merged, but follow-up will implement both sources side by side and the Mix.Task for refreshing data. 

I think refactoring a large portion of the sources module tests once everything is implemented would be a good idea (hence the follow up PRs).

### CODE REVIEW

File: apps/breethe/lib/breethe/sources/eea.ex [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      "GE",
+      "MK",
+      "PL",
+      "AT",
+      "CH",
+      "EE"
+    ]
+
+  # @parameters ["PM10", "PM2.5", "SO2", "NO2", "O3", "CO"]
+  def parameters(), do: ["SO2", "NO2", "O3", "CO"]
```
Comment (marcoow) [inline apps/breethe/lib/breethe/sources/eea.ex]: Why no `@parameters`?

File: apps/breethe/lib/breethe/sources/eea.ex [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+  def get_data() do
+    for country <- country_codes(), param <- parameters() do
+      country
```
Comment (marcoow) [inline apps/breethe/lib/breethe/sources/eea.ex]: Can we run each one in its own `Task`? Or maybe use https://github.com/dashbitco/broadway ?

File: apps/breethe/lib/breethe/sources/eea.ex (hunk 1 above)
Comment (niklaslong) [reply apps/breethe/lib/breethe/sources/eea.ex]: I exposed the `params` and the `country_codes` in functions to avoid code duplication in the tests, not sure if this is a good approach though?

File: apps/breethe/lib/breethe/sources/eea.ex (hunk 2 above)
Comment (niklaslong) [reply apps/breethe/lib/breethe/sources/eea.ex]: Yep, my intention was to implement this in a follow-up (also because I'll be refactoring a few things including how we manage asnyc tasks for requests)?

**Comment by niklaslong** [comment]
@marcoow I've update mix.lock as well since some updates were done since. Any thoughts on merging this as is and looking into async in the follow-up when I'll connect everything?

File: apps/breethe/lib/breethe/sources/eea.ex (hunk 1 above)
Comment (marcoow) [reply apps/breethe/lib/breethe/sources/eea.ex]: Yeah, makes sense - I think a more common idiom is

```elixir
@parameters ["PM10", "PM2.5", "SO2", "NO2", "O3", "CO"]
def parameters(), do: @parameters
```

though?

Comment (niklaslong) [reply apps/breethe/lib/breethe/sources/eea.ex]: fixed

File: apps/breethe/test/breethe/sources/eea/csv_test.exs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+  @sample_valid_data "network_countrycode,network_localid,network_name,network_namespace,network_timezone,pollutant,samplingpoint_localid,samplingpoint_namespace,samplingpoint_x,samplingpoint_y,coordsys,station_code,station_localid,station_name,station_namespace,value_datetime_begin,value_datetime_end,value_datetime_inserted,value_datetime_updated,value_numeric,value_validity,value_verification,station_altitude,value_unit\r\nGB,NET-GB001A,National air network,GB.GB-ExEA.AQ,http://dd.eionet.europa.eu/vocabulary/aq/timezone/UTC+02,SO2,SPO-BG0071A_00001_100,GB.GB-ExEA.AQ,00.72,34.65975799999999,EPSG:4979,GB0071A,STA-GB0071A,London,GB.GB-ExEA.AQ,2019-11-10 23:00:00+01:00,2019-11-11 00:00:00+01:00,2019-11-11 01:29:21+01:00,+01:00,3,-1,1,27,ug/m3\r\n"
```
Comment (marcoow) [inline apps/breethe/test/breethe/sources/eea/csv_test.exs]: Might be easier to have these in fixture files? It's a bit hard to follow here what the actual data is in a super long string like this.

File: apps/breethe/lib/breethe/sources/eea.ex [hunk 4]
```diff
@@ -0,0 +1,65 @@
+defmodule Breethe.Sources.EEA do
+  alias __MODULE__.CSV
+
+  # @parameters ["PM10", "PM2.5", "SO2", "NO2", "O3", "CO"]
+  @parameters ["SO2", "NO2", "O3", "CO"]
+
+  def country_codes(),
```
Comment (marcoow) [inline apps/breethe/lib/breethe/sources/eea.ex]: I'd change this to use the same pattern as for `@parrameters` with `def country_codes(), do: @country_codes` for consistency.

File: apps/breethe/test/breethe/sources/eea/csv_test.exs (hunk 3 above)
Comment (niklaslong) [reply apps/breethe/test/breethe/sources/eea/csv_test.exs]: Wouldn't that just move the problem from one file to another though?

Comment (niklaslong) [reply apps/breethe/test/breethe/sources/eea/csv_test.exs]: Perhaps programmatically creating the data would be a better solution?

File: apps/breethe/lib/breethe/sources/eea.ex (hunk 4 above)
Comment (niklaslong) [reply apps/breethe/lib/breethe/sources/eea.ex]: fixed

File: apps/breethe/test/breethe/sources/eea/csv_test.exs (hunk 3 above)
Comment (marcoow) [reply apps/breethe/test/breethe/sources/eea/csv_test.exs]: Hm, I think actually the best solutions would be keeping it here but properly formatting it as a multiline string: https://elixir-examples.github.io/examples/multiline-strings-heredocs

Comment (niklaslong) [reply apps/breethe/test/breethe/sources/eea/csv_test.exs]: fixed - this is the simplest I could do without implementing a custom sigil

**Comment by niklaslong** [comment]
The resulting string of what you wrote above would be:
```
"col1,col2,col3\nval1,val2,val3\n" 
```
We could then process the string to remove the added `"\n"` characters. The issue with this however is that the csv data already contains newlines and carriage returns. It's a little tricky to work out which ones to remove and which ones to keep as there is no way of telling the new ones from the original ones. Which is why I tried string concatenation (which would not introduce any new characters and make the data more readable):
```
"col1," <>
"col2," <>
...
"val1," <>
"val2"
```
This worked well, but I then thought I could refactor using lists (and avoid a lot of duplication) to what is currently in the PR. 

I understand that there's a bit of logic involved in putting the data together, which isn't a good thing. However it seems to me that readability will come at that cost in this case? I'm not sure what the best solution is, but I don't think we can achieve what you suggested?

**Comment by marcoow** [comment]
I think you're over-complicating things. You should be able to just copy the data you want to have from the actual CSV file and put it between `"""` in the `.ex` file – `"""` preserving the newlines is actually what we want here, isn't it?

File: apps/breethe/test/breethe/sources/eea/csv_test.exs [hunk 5]
```diff
@@ -0,0 +1,44 @@
+defmodule Breethe.Sources.EEA.CSVTest do
+  use Breethe.DataCase
+
+  alias Breethe.Sources.EEA.CSV
+
+  @sample_valid_data """
+  network_countrycode,network_localid,network_name,network_namespace,network_timezone,pollutant,samplingpoint_localid,samplingpoint_namespace,samplingpoint_x,samplingpoint_y,coordsys,station_code,station_localid,station_name,station_namespace,value_datetime_begin,value_datetime_end,value_datetime_inserted,value_datetime_updated,value_numeric,value_validity,value_verification,station_altitude,value_unit\r
+  GB,NET-GB001A,National air network,GB.GB-ExEA.AQ,http://dd.eionet.europa.eu/vocabulary/aq/timezone/UTC+02,SO2,SPO-BG0071A_00001_100,GB.GB-ExEA.AQ,00.72,34.65975799999999,EPSG:4979,GB0071A,STA-GB0071A,London,GB.GB-ExEA.AQ,2019-11-10 23:00:00+01:00,2019-11-11 00:00:00+01:00,2019-11-11 01:29:21+01:00,+01:00,3,-1,1,27,ug/m3\r
+  """
```
Comment (niklaslong) [inline apps/breethe/test/breethe/sources/eea/csv_test.exs]: Is this what you meant?

Comment (marcoow) [reply apps/breethe/test/breethe/sources/eea/csv_test.exs]: yes exactly – at least this is exactly the same as what you'd have in an actual CSV file.
